### PR TITLE
Batch close-indices cluster state updates

### PR DESCRIPTION
**Pull request by joegallo**
Related to https://github.com/elastic/elasticsearch/issues/81627 and https://github.com/elastic/elasticsearch/pull/83432, followup to https://github.com/elastic/elasticsearch/pull/83760.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java [hunk 1]
```diff
@@ -1023,24 +1088,7 @@ public static ClusterBlock createUUIDBasedBlock(ClusterBlock clusterBlock) {
...
-            IndexMetadataVerifier indexMetadataVerifier,
-            IndicesService indicesService,
-            ShardLimitValidator shardLimitValidator
-        ) {
-            this.allocationService = allocationService;
-            this.indexMetadataVerifier = indexMetadataVerifier;
-            this.indicesService = indicesService;
-            this.shardLimitValidator = shardLimitValidator;
-        }
+    private class OpenIndicesExecutor implements ClusterStateTaskExecutor<OpenIndicesTask> {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java:1091]: Slight preference to keep this (and other executors) as static classes that hold onto the services they need directly.

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java [hunk 2]
```diff
@@ -147,114 +157,170 @@ public MetadataIndexStateService(
...
+                    Map<Index, ClusterBlock> blockedIndices = task.blockedIndices;
+                    Map<Index, IndexResult> verifyResults = task.verifyResults;
+                    final List<IndexResult> indices = new ArrayList<>();
+
+                    Tuple<ClusterState, Collection<IndexResult>> closingResult = closeRoutingTable(state, blockedIndices, verifyResults);
+                    assert verifyResults.size() == closingResult.v2().size();
+                    indices.addAll(closingResult.v2());
+                    state = closingResult.v1();
+
+                    builder.success(task, new ActionListener<>() {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java:256]: Suggest `task.listener().delegateFailure(...)` here.

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java [hunk 3]
```diff
@@ -147,114 +157,170 @@ public MetadataIndexStateService(
...
+
+            for (CloseIndicesTask task : tasks) {
+                try {
+                    Map<Index, ClusterBlock> blockedIndices = task.blockedIndices;
+                    Map<Index, IndexResult> verifyResults = task.verifyResults;
+                    final List<IndexResult> indices = new ArrayList<>();
+
+                    Tuple<ClusterState, Collection<IndexResult>> closingResult = closeRoutingTable(state, blockedIndices, verifyResults);
+                    assert verifyResults.size() == closingResult.v2().size();
+                    indices.addAll(closingResult.v2());
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java:253]: Do we need to copy `closingResult.v2()` to a new list here? Think we could just use it directly in the publish listener.

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java [hunk 4]
```diff
@@ -147,114 +157,170 @@ public MetadataIndexStateService(
...
+                        public void onResponse(ClusterState clusterState) {
+                            if (blockedIndices.isEmpty()) {
+                                task.listener.onResponse(new CloseIndexResponse(true, false, Collections.emptyList()));
+                            } else {
+                                threadPool.executor(ThreadPool.Names.MANAGEMENT)
+                                    .execute(
+                                        new WaitForClosedBlocksApplied(
+                                            blockedIndices,
+                                            task.request,
+                                            ActionListener.wrap(
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java:195]: and here, `task.listener().delegateFailure(...)`.

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java [hunk 5]
```diff
@@ -147,114 +157,170 @@ public MetadataIndexStateService(
...
+        public ClusterTasksResult<AddBlocksToCloseTask> execute(ClusterState currentState, List<AddBlocksToCloseTask> tasks)
+            throws Exception {
+            ClusterTasksResult.Builder<AddBlocksToCloseTask> builder = ClusterTasksResult.builder();
+            ClusterState state = currentState;
+
+            for (AddBlocksToCloseTask task : tasks) {
+                try {
+                    final Map<Index, ClusterBlock> blockedIndices = new HashMap<>(task.request.indices().length);
+                    state = addIndexClosedBlocks(task.request.indices(), blockedIndices, state);
+                    builder.success(task, new ActionListener<>() {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java:184]: and here, `task.listener().delegateFailure(...)`.

**Review by DaveCTurner — COMMENTED** [review]
Looks good, I left a few nits but seems structurally sound to me.

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java [hunk 6]
```diff
@@ -147,114 +157,170 @@ public MetadataIndexStateService(
...
+                        @Override
+                        public void onResponse(ClusterState clusterState) {
+                            if (blockedIndices.isEmpty()) {
+                                task.listener.onResponse(new CloseIndexResponse(true, false, Collections.emptyList()));
+                            } else {
+                                threadPool.executor(ThreadPool.Names.MANAGEMENT)
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java:190]: I don't see a great reason to fork here, we really just trigger a bunch of transport actions that themselves fork to `MANAGEMENT` again. But then again this is sending `O(#shards)` transport actions in parallel, which could be lots, and we know that can cause problems. No change needed here in this PR but we might want to do something about that in future perhaps.

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java (hunk 2 above)
Comment (joegallo) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java]: 053c762cfcdbb77c79a1589a6e957cd17dcef423

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java (hunk 1 above)
Comment (joegallo) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java:1074]: Ehhhh, this is a bit of a pain -- `private class WaitForClosedBlocksApplied` / ` private class WaitForBlocksApplied`, and the former is invoked by `AddBlocksToCloseExecutor`, so I'd have to walk those two back, too.

I'm really with you on disliking (non-static) inner classes, they feel a bit too ['more magic'](http://www.catb.org/jargon/html/magic-story.html) for me, but in this case they're all private anyway, and the juggling of the fields across them all just seems like noise.

That said, this isn't 'no', it's just 'ehhh, this is ever so slightly harder than you might have thought' -- feel free to say something like "yeah, that's cool, please make the change anyway" and I will happily do it.

Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java:1074]: My preference is only slight, I'll defer to your judgement.

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java [hunk 7]
```diff
@@ -147,114 +157,153 @@ public MetadataIndexStateService(
...
+            ClusterTasksResult.Builder<AddBlocksToCloseTask> builder = ClusterTasksResult.builder();
+            ClusterState state = currentState;
+
+            for (AddBlocksToCloseTask task : tasks) {
+                try {
+                    final Map<Index, ClusterBlock> blockedIndices = new HashMap<>(task.request.indices().length);
+                    state = addIndexClosedBlocks(task.request.indices(), blockedIndices, state);
+                    builder.success(task, task.listener.delegateFailure((l1, clusterState) -> {
+                        if (blockedIndices.isEmpty()) {
+                            task.listener.onResponse(new CloseIndexResponse(true, false, Collections.emptyList()));
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java:186]: Should really use the delegated listener here, although secretly it's the listener we first thought of:

```suggestion
                            l1.onResponse(new CloseIndexResponse(true, false, Collections.emptyList()));
```

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java [hunk 8]
```diff
@@ -147,114 +157,153 @@ public MetadataIndexStateService(
...
+                        } else {
+                            threadPool.executor(ThreadPool.Names.MANAGEMENT)
+                                .execute(
+                                    new WaitForClosedBlocksApplied(
+                                        blockedIndices,
+                                        task.request,
+                                        task.listener.delegateFailure((l2, verifyResults) -> {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java:193]:
```suggestion
                                        l1.delegateFailure((l2, verifyResults) -> {
```

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java [hunk 9]
```diff
@@ -147,114 +157,153 @@ public MetadataIndexStateService(
...
+                                            clusterService.submitStateUpdateTask(
+                                                "close-indices",
+                                                new CloseIndicesTask(task.request, blockedIndices, verifyResults, task.listener),
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java:196]:
```suggestion
                                                new CloseIndicesTask(task.request, blockedIndices, verifyResults, l2),
```

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java [hunk 10]
```diff
@@ -147,114 +157,153 @@ public MetadataIndexStateService(
...
+                                            "[{}] indices closed, but the operation timed out while "
+                                                + "waiting for enough shards to be started.",
+                                            Arrays.toString(waitForIndices)
+                                        );
+                                    }
+                                    // acknowledged maybe be false but some indices may have been correctly closed,
+                                    // so we maintain a kind of coherency by overriding the shardsAcknowledged value
+                                    // (see ShardsAcknowledgedResponse constructor)
+                                    boolean shardsAcked = acknowledged ? shardsAcknowledged : false;
+                                    task.listener.onResponse(new CloseIndexResponse(acknowledged, shardsAcked, indices));
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java:272]:
```suggestion
                                    l.onResponse(new CloseIndexResponse(acknowledged, shardsAcked, indices));
```

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java [hunk 11]
```diff
@@ -147,114 +157,153 @@ public MetadataIndexStateService(
...
+                                },
+                                task.listener::onFailure
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java:274]:
```suggestion
                                l::onFailure
```

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java [hunk 12]
```diff
@@ -147,114 +157,153 @@ public MetadataIndexStateService(
...
                             );
-                    }
+                        } else {
+                            task.listener.onResponse(new CloseIndexResponse(acknowledged, false, indices));
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java:277]:
```suggestion
                            l.onResponse(new CloseIndexResponse(acknowledged, false, indices));
```

**Review by DaveCTurner — COMMENTED** [review]
Nits

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java (hunk 8 above)
Comment (joegallo) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java]: 1485315cc6ef89c399c142daec05140b8a075794

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java (hunk 9 above)
Comment (joegallo) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java]: 1485315cc6ef89c399c142daec05140b8a075794

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java (hunk 10 above)
Comment (joegallo) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java]: 1485315cc6ef89c399c142daec05140b8a075794

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java (hunk 11 above)
Comment (joegallo) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java]: 1485315cc6ef89c399c142daec05140b8a075794

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java (hunk 12 above)
Comment (joegallo) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java]: 1485315cc6ef89c399c142daec05140b8a075794

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java (hunk 3 above)
Comment (joegallo) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java]: d3445d2a27bad2477845c81d794c205a80f00595 and then d64a1037c7dac962bc309ea49fed212ffa0e7854

**Review by DaveCTurner — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java [hunk 13]
```diff
@@ -460,7 +505,7 @@ public ClusterState execute(final ClusterState currentState) {
                 public void clusterStateProcessed(final ClusterState oldState, final ClusterState newState) {
                     if (oldState == newState) {
                         assert blockedIndices.isEmpty() : "List of blocked indices is not empty but cluster state wasn't changed";
-                        listener.onResponse(new AddIndexBlockResponse(true, false, Collections.emptyList()));
+                        listener.onResponse(new AddIndexBlockResponse(true, false, List.of()));
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java:508]: NIT: same here, if we're changing this line anyway we could make the empty response a constant

**Review by original-brownbear — APPROVED** [review]
LGTM (nits really can be ignored, just had to let them out :))

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java [hunk 14]
```diff
@@ -147,114 +153,153 @@ public MetadataIndexStateService(
...
+            ClusterTasksResult.Builder<AddBlocksToCloseTask> builder = ClusterTasksResult.builder();
+            ClusterState state = currentState;
+
+            for (AddBlocksToCloseTask task : tasks) {
+                try {
+                    final Map<Index, ClusterBlock> blockedIndices = new HashMap<>(task.request.indices().length);
+                    state = addIndexClosedBlocks(task.request.indices(), blockedIndices, state);
+                    builder.success(task, task.listener.delegateFailure((delegate1, clusterState) -> {
+                        if (blockedIndices.isEmpty()) {
+                            delegate1.onResponse(new CloseIndexResponse(true, false, List.of()));
```
[the remark this answers is not in the document]
Comment (joegallo) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java]: aa32395

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java (hunk 13 above)
Comment (joegallo) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java]: aa32395

**Pull request merged**